llvm/labs/lab2/vetoshnikova_ekaterina/InstrumentFunctions.cpp

```diff
@@ -0,0 +1,84 @@
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace {
+
+struct InstrumentFunctions : llvm::PassInfoMixin<InstrumentFunctions> {
+
+  llvm::PreservedAnalyses run(llvm::Function &F,
+                              llvm::FunctionAnalysisManager &AM) {
+
+    llvm::LLVMContext &context = F.getContext();
+
+    llvm::IRBuilder<> builder(context);
+
+    auto module = F.getParent();
+
+    llvm::FunctionType *type =
+        llvm::FunctionType::get(llvm::Type::getVoidTy(context), false);
+
+    llvm::FunctionCallee instrumentStart =
+        module->getOrInsertFunction("instrument_start", type);
+
+    llvm::FunctionCallee instrumentEnd =
+        module->getOrInsertFunction("instrument_end", type);
+
+    bool fStart = false;
+    bool fEnd = false;
+
+    llvm::CallInst *CInst;
+
+    for (llvm::BasicBlock &b : F) {
+      for (auto &instr : b) {
+        if ((CInst = llvm::dyn_cast<llvm::CallInst>(&instr)) != NULL) {
+          CInst = llvm::cast<llvm::CallInst>(&instr);
+          if (CInst->getCalledFunction() == instrumentStart.getCallee()) {
```

Comment:
Please check instrumentStart/End availability via checking users of these functions declarations. It is faster in common.